# Sesiunea 14 – NumPy Avansat
_Notebook de exerciții (fără soluții)._

## Exercițiu 1 – Broadcasting cu bias
Creează un array `img` 3x3 și un vector `bias`, apoi aplică biasul pe fiecare rând folosind broadcasting.

Exemplu:
```python
img = np.array([[10,20,30],[40,50,60],[70,80,90]])
bias = np.array([1, 0, -1])
```
Afișează rezultatul lui `img + bias`.

In [3]:
import numpy as np

img = np.array([[1,2,3],[4,5,6],[7,8,9]])
bias = np.array([4,3,2])

print(img+bias)

[[ 5  5  5]
 [ 8  8  8]
 [11 11 11]]


## Exercițiu 2 – Indexare logică
Creează un array 1D de valori numerice. Folosește o mască logică pentru a extrage doar elementele care respectă o anumită condiție (de exemplu, valori mai mari de 25).

In [5]:
import numpy as np
arr = np.array([5, 14, 90, 35, 109, 45])
mask = arr > 25
print(mask)
numbers_over_25 = arr[mask]
print('Numere peste 25: ', numbers_over_25)

[False False  True  True  True  True]
Numere peste 25:  [ 90  35 109  45]


## Exercițiu 3 – Funcții matematice și agregări
Creează un vector `x` folosind `np.linspace(0, 10, 100)`. Calculează `sin(x)`, apoi media și deviația standard.

În plus, aplică operații de agregare pe o matrice 2D folosind `axis=0` și `axis=1`.

In [6]:
#exercise 3 - Funcții matematice și agregări

import numpy as np

#np.linspace() creates evenly spaced numbers between a start and end point
x = np.linspace(0,10,100) #asta nu da numerele de la 1 la 100, ci 100 de numere dintre 0 si 100, equally spaced
x_style = print('x is:', x,)

#np.sin() calculates the sine of each element
sin_x = print('sin of x is:', np.sin(x))

media_x = print('Media of x numbers:', np.mean(x))

#deviatia(deviation) masura cat de despartite sunt numerele intre ele
#High std = numbers are far apart
#Low std = numbers are close together

deviatia = print('Deviatia este:', np.std(x))

#aplică operații de agregare pe o matrice 2D folosind axis=0 și axis=1
#axis=0 = operate down the columns (across rows)
#axis=1 = operate across the rows (along columns)

matrice_2d = ([[1,2,3], 
               [4,5,6],
               [7,8,9]])
print('axis0:', np.sum(matrice_2d, axis=0)) #sums across rows
print('axis1:', np.sum(matrice_2d, axis = 1))  #sums across columns

x is: [ 0.          0.1010101   0.2020202   0.3030303   0.4040404   0.50505051
  0.60606061  0.70707071  0.80808081  0.90909091  1.01010101  1.11111111
  1.21212121  1.31313131  1.41414141  1.51515152  1.61616162  1.71717172
  1.81818182  1.91919192  2.02020202  2.12121212  2.22222222  2.32323232
  2.42424242  2.52525253  2.62626263  2.72727273  2.82828283  2.92929293
  3.03030303  3.13131313  3.23232323  3.33333333  3.43434343  3.53535354
  3.63636364  3.73737374  3.83838384  3.93939394  4.04040404  4.14141414
  4.24242424  4.34343434  4.44444444  4.54545455  4.64646465  4.74747475
  4.84848485  4.94949495  5.05050505  5.15151515  5.25252525  5.35353535
  5.45454545  5.55555556  5.65656566  5.75757576  5.85858586  5.95959596
  6.06060606  6.16161616  6.26262626  6.36363636  6.46464646  6.56565657
  6.66666667  6.76767677  6.86868687  6.96969697  7.07070707  7.17171717
  7.27272727  7.37373737  7.47474747  7.57575758  7.67676768  7.77777778
  7.87878788  7.97979798  8.08080808  8.18181

## Exercițiu 4 – Normalizare pe axă
Normalizează o matrice 2x2 folosind formula:

\[ norm = (m - media) / std \]

unde media și deviația standard sunt calculate pe coloane (`axis=0`).

In [7]:
#Normalization (also called standardization) transforms data so that:
#The mean becomes 0
#The standard deviation becomes 1

#formula: norm = (m - mean) / std = 
#m = my original matrix
#mean = media per column (axis=0)
#std = standard deviation per column (axis=0)

import numpy as np
matrix_2x2 = np.array([[1,2], 
               [3,4]])
media_matrix_2x2 = np.mean(matrix_2x2, axis=0)
standard_deviation_per_column = np.std(matrix_2x2, axis=0)

normalized = (matrix_2x2 - media_matrix_2x2) / standard_deviation_per_column
print(normalized)



[[-1. -1.]
 [ 1.  1.]]


## Exercițiu 5 – Manipulare avansată de array-uri
Creează un vector `v` cu 12 elemente, transformă-l într-o matrice 3x4. Aplică:
- `np.hsplit()` pentru a-l împărți pe coloane
- `np.vstack()` pentru a concatena două copii ale matricei pe verticală.

In [9]:
#exercitiu 5 - Manipulare avansată de array-uri
import numpy as np
v = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12])  # vector 1D cu 12 elemente

v_reshape = v.reshape(3,4)
print('matrix:\n', v_reshape) 

#reshape = schimbarea formey unui array fara a schimba datele
#hsplit = horizontal split (impartire orizontala pe coloane, imparte matricea in bucati vertical separate pe coloane)
print('splitul:\n', np.hsplit(v_reshape, 2)) #aici trebuie sa pun un numar care se imparte la 4 fiindac am 4 coloane. cand pun 3 de exemply, nu merge sa impart 4 la 3. deci 2 e okay!


#vstack = vertical stack, pune array-uri unul sub celaalt pe verticala
print('stack-ul:\n', np.vstack([v_reshape, v_reshape]))

matrix:
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]
splitul:
 [array([[ 1,  2],
       [ 5,  6],
       [ 9, 10]]), array([[ 3,  4],
       [ 7,  8],
       [11, 12]])]
stack-ul:
 [[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]


## Exercițiu 6 – Vectorizare vs For-loop
Generează un milion de numere și calculează pătratele lor:
- o dată folosind un for-loop
- o dată folosind operații vectorizate cu NumPy
Compară timpul de execuție între cele două metode.

In [10]:
numere_random = np.arange(1000000)
#print(numere_random)

#metoda 1 for loop
import time
start_time_forloop = time.time()
for_loop = []
for fiecare_numar in numere_random:
    patratul_numarului_loop = fiecare_numar ** 2
    for_loop.append(patratul_numarului_loop)


end_time_forloop = time.time()
end_loop = end_time_forloop - start_time_forloop
print('Timpul prin for loop este:', end_loop)

#metoda 2 vectorizare
start_time_vectorizare = time.time()
patrat_loop = (numere_random ** 2)

end_time_vectorizare = time.time()

end_vectorizare = end_time_vectorizare - start_time_vectorizare
print('Timpul prin vectorizare este: ', end_vectorizare)

print('For-loop ul este de', end_loop / end_vectorizare, 'ori mai lent decat vectorizarea')



Timpul prin for loop este: 0.3195772171020508
Timpul prin vectorizare este:  0.013905525207519531
For-loop ul este de 22.982031410740003 ori mai lent decat vectorizarea


## Exercițiu 7 – Distanță euclidiană vectorizată
Calculează distanța euclidiană dintre doi vectori `p1` și `p2` fără a folosi bucle.

\[ dist = \sqrt{\sum (p1 - p2)^2} \]

## Exercițiu 8 – Big Arrays și filtrare
Generează un array de 10 milioane de numere întregi aleatoare între 1 și 20000.
Selectează doar valorile mai mari de 10000 și calculează media acestora.

## Exercițiu 9 – Exerciții practice
- Generează 1 milion de numere aleatoare și calculează suma celor divizibile cu 7.
- Normalizează o matrice 1000x1000.
- Simulează o imagine RGB 1920x1080 și înmulțește fiecare pixel cu 1.5.
- Creează o matrice 3D și calculează suma pe axa Z (`axis=2`).